[WellSQL Migration] Migrate `WCProductVariationModel` to Room

ParaskP7 opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: [AINFRA-124](https://linear.app/a8c/issue/AINFRA-124)

FYI: This is a feature branch of two PRs ([context](https://github.com/woocommerce/woocommerce-android/pull/13972#issuecomment-2862684257)):
- #13972
- #13982

---

### Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->

This PR migrates `WCProductVariationModel` to Room.

FYI: For more info see [this](https://github.com/woocommerce/woocommerce-android/pull/13972) PR's description as well.

---

### Testing information
<!-- This is your opportunity to break out individual scenarios that need testing (when necessary) and/or include a checklist for the reviewer to go through. Consider documenting the following from your own completed testing: devices used, alternate workflows, edge cases, affected areas, critical flows, areas not tested, and any remaining unknowns. Provide feedback on this new section of the PR template through Sept 30, 2024 to Apps Quality; additional context here: p91TBi-b8z-p2#comment-12036 -->

Smoke test any feature that uses `WCProductVariationModel`.

FYI: For more info see [this](https://github.com/woocommerce/woocommerce-android/pull/13982) PR's testing information as well.

---

### Images/gif
<!-- Include before and after images or gifs when appropriate. -->

![image](https://github.com/user-attachments/assets/d1a2f30b-cb95-40fc-9820-c22c0f9fde92)

- [x] I have considered if this change warrants release notes and have added them to `RELEASE-NOTES.txt` if necessary. Use the "[Internal]" label for non-user-facing changes.

---

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:

- [ ] The PR is small and has a clear, single focus, or a valid explanation is provided in the description. If needed, please request to split it into smaller PRs.
- [ ] Ensure Adequate Unit Test Coverage: The changes are reasonably covered by unit tests or an explanation is provided in the PR description.
- [ ] Manual Testing: The author listed all the tests they ran, including smoke tests when needed (e.g., for refactorings). The reviewer confirmed that the PR works as expected on big (tablet) and small (phone) in case of UI changes, and no regressions are added.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

wzieba committed:
Migrate `WCProductVariationModel` to Room

`:libs:fluxc-plugin:assembleDebug` passes

wzieba committed:
Migrate `getVariations` and `observeVariations` to Room's DAO

wzieba committed:
Migrate `getVariation` to Room's DAO

wzieba committed:
Migrate `upsertVariation`(s) to Room's DAO

wzieba committed:
Migrate `deleteVariation` to Room's DAO

wzieba committed:
Remove unused observability wrappers for variations in `ProductSqlUtils`

wzieba committed:
Adjust `WooCommerce` module to WCProductVariationModel changes: update to app model mapper, make query methods suspendable

wzieba committed:
Fix operations on ProductVariationsDao by passing **local** site id

To reduce chance of similar mistake in the future, I propose using `LocalId/RemoteId` wrappers.

wzieba committed:
Fix `ProductTestUtils` from testFixtures

To fix lint task

wzieba committed:
Fix building `fluxc-plugin` tests

2 tests are failing right now

wzieba committed:
Ignore variations order in `test that variation insert emits a flow even`

Room and WellSql have different defaults for sort. In this test, the order doesn't matter as all items have `menuOrder=0`

wzieba committed:
Fix WCProductStore unit tests

Don't use incorrect `$original.id` for description: instead, use simple string to avoid any confusion.

Close database on test end

wzieba committed:
Apply code style

wzieba committed:
Make the `testJalapenoDebugUnitTest` run

wzieba committed:
Fix `WooCommerce` unit tests by mocking suspendable functions to return default values

Previously, functions like `VariationRepository#getProductVariationList` or `ProductDetailRepository#getCachedVariationCount` were *not* suspend. This allowed Mockito to return default values for both of the methods, respectively: empty list and 0.

After recent changed, both of this functions were refactored to suspend, which caused Mockito to return `null` for them (see: https://github.com/mockito/mockito-kotlin/issues/342)

This comment brings back the previous behavior by mocking mentioned methods to return default values.

wzieba committed:
Drop `WCProductVariationModel` from WellSql

wzieba committed:
Make `ProductVariationsDao` internal

wzieba committed:
Merge branch 'trunk' into migrate_product_variation_to_room

wzieba committed:
Merge branch 'feature/product_variation_to_room' into migrate_product_variation_to_room

wzieba committed:
Merge branch 'migrate_product_variation_to_room' into migrate_product_variation_to_room_tests

ParaskP7 committed:
Merge pull request #13972 from woocommerce/migrate_product_variation_to_room

[1/2] `WCProductVariationModel` to Room: working app

ParaskP7 committed:
Merge pull request #13982 from woocommerce/migrate_product_variation_to_room_tests

[2/2] `WCProductVariationModel` to Room: unit tests adjustments

ParaskP7 committed:
Merge branch 'trunk' of github.com:woocommerce/woocommerce-android into feature/product_variation_to_room

dangermattic commented:
<!--
  0 Errors
  2 Warnings: This PR is larger than 300 lin...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          2 Warnings
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 300 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/woocommerce/woocommerce-android/milestone/258">22.4</a>. This milestone is due in less than 2 days.<br />
Please make sure to get it merged by then or assign it to a milestone with a later deadline.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

ParaskP7 committed:
Fix: Replace payload with result on delete variations for product

It looks like as part of 1df49aa69f4f15d1bc702f33746cc7003a62e790 an
oversight occurred where, instead of 'result.remoteProductId',
'payload.remoteProductId' got used.

I am not sure if that was deliberate or not, and since there was no
specific comment about that, it is better to treat that as an oversight.
This commit serves as an explicit nudge to reviewers, and the original
auhtor, to verify this change with that extra 'fix' commit.

ParaskP7 committed:
Refactor: Use local/remote instead of int/long ids for product variation

This follows the same pattern applied to 'WCProductModel' and it seems
that this pattern is the one mostly used. This pattern is similar to
using a 'value' class to create a more domain-specific type instead of
primitive types to help with data class construction clarity.

Also, related to: 316ef5fae66bb639beaf1e47342e6c26c0875953

ParaskP7 committed:
Merge branch 'trunk' of github.com:woocommerce/woocommerce-android into feature/product_variation_to_room

ParaskP7 commented:
👋 @malinajirka , I am adding you as an extra reviewer to this, additional to myself, me taking this to the finish line for @wzieba (who is currently AFK). I am doing that because I want someone from the product team to do a final review on this, test this as thoroughly as possible and signing this change for merging.

With that, I want to give you some extra hints on where I would love you to put some extra focus:
1. This 9f715a424a6dac967455ed35c9bd1bfd873f2fa1 commit and more specifically the new `primaryKeys` constraint (see [comment](https://github.com/woocommerce/woocommerce-android/pull/13972#discussion_r2063760434)). 👀 
2. This 9d12e9d8a91e256bada57535d4e56322b212c223 commit and more specifically the new `attributesToJson()` function (see [discussion](https://github.com/woocommerce/woocommerce-android/pull/13972#discussion_r2063758138)). 👀
3. This extra a440b6bc96f596ce0de10bd6a6ba029b5d13cc52 commit (of mine) that is supposed to be fixing an oversight. 🤔
4. This extra 375d22af9d4d3c014db0c2dead3494aa044e0c19 commit (of mine) that refactors the entity to utilize `LocalId`/`RemoteId` in favor of `Int`/`Long`.
    - PS: I am actually wondering what's the best practice you all follow atm on this. 🤔
6. This testing related [comment](https://github.com/woocommerce/woocommerce-android/pull/13982#issuecomment-2864004331) of mine and whether this problem happens on your side too. 🧪

💭 Last but not least, and maybe more importantly, I wanted to clarify the behavior of this migration. I want to clarify that with this kind of migrations we are okay dropping an existing table ([wp-fluxc.WCProductVariationModel](https://github.com/woocommerce/woocommerce-android/pull/14027/files#diff-165cc8670b9fdfb667171d66d638e2a775233ea2d5bf51d3043824f3235b3ab6R2130-R2132)), along with its entries, and recreating the table ([wc-android-database.ProductVariationEntity](https://github.com/woocommerce/woocommerce-android/pull/14027/files#diff-c1dc6c021892393397fd1b7b7c26e9a85a1eca0081730fa066a0c537348042c6R128)), but without any entries (empty). With such a migration, all previous table entries will be gone and added back (partially or in full) only when a user navigates to that specific screen.
- Firstly, I guess that since #13935 has been already merged, this is an acceptable migration, and actually, maybe the only way to do such a migration.
- Nevertheless, I just want to take this opportunity to clarify this, and furthermore, to ask if you can foresee any trouble with such a migration (me thinking of online/offline sync, etc).

Once more, thank you so much in advance for all the help and insights on this! 🙏

malinajirka commented:
Thanks for working on this @ParaskP7!

I'll also invite @hichamboushaba and @kidinov since they reviewed the previous PRs so likely know what to look for. Moreover, I'm wrapping up a project with a target date set to next friday and I'm a bit behind, so it might take me a bit before I find time for this review (I'd estimate it to Wednesday the earliest).

ParaskP7 commented:
Thanks for all the help @malinajirka , the heads-up on this and for inviting others to review this, much appreciated! 🙇

ParaskP7 committed:
Merge branch 'trunk' of github.com:woocommerce/woocommerce-android into feature/product_variation_to_room

ParaskP7 commented:
👋 team (@malinajirka @hichamboushaba @kidinov), just a friendly ping as a reminder on this, thank you! 🙏

malinajirka commented:
> This extra https://github.com/woocommerce/woocommerce-android/commit/a440b6bc96f596ce0de10bd6a6ba029b5d13cc52 commit (of mine) that is supposed to be fixing an oversight. 🤔

I think this change is fine, but just to double-check my understanding. I believe the remoteProductId between payload and result is always identical, right? Or have you fixed it based on some bug you discovered?

> This extra https://github.com/woocommerce/woocommerce-android/commit/375d22af9d4d3c014db0c2dead3494aa044e0c19 commit (of mine) that refactors the entity to utilize LocalId/RemoteId in favor of Int/Long.
> PS: I am actually wondering what's the best practice you all follow atm on this. 🤔

Thanks, that's a nice improvement! The wrappers were introduced as part of the introduction of List Management component in WPAndroid (for posts) and later in WCAndroid (for orders). I'm not aware of any guidelines/best-practices regarding the usage - for models that are using the list management component, this is required, for other models, it's nice to have.

> This testing related https://github.com/woocommerce/woocommerce-android/pull/13982#issuecomment-2864004331 of mine and whether this problem happens on your side too. 🧪

I can confirm the byteBuddy exception is happening on my side as well 🤔. Having said that, it's happening even when I run the command on `trunk`, so I don't think it's related this PR.

> 💭 Last but not least, and maybe more importantly, I wanted to clarify the behavior of this migration. I want to clarify that with this kind of migrations we are okay dropping an existing table ([wp-fluxc.WCProductVariationModel](https://github.com/woocommerce/woocommerce-android/pull/14027/files#diff-165cc8670b9fdfb667171d66d638e2a775233ea2d5bf51d3043824f3235b3ab6R2130-R2132)), along with its entries, and recreating the table ([wc-android-database.ProductVariationEntity](https://github.com/woocommerce/woocommerce-android/pull/14027/files#diff-c1dc6c021892393397fd1b7b7c26e9a85a1eca0081730fa066a0c537348042c6R128)), but without any entries (empty). With such a migration, all previous table entries will be gone and added back (partially or in full) only when a user navigates to that specific screen.

Editing products and variations is supported in the app, but we do not support offline => the user needs to explicitly save the changes while being online => AFAIK it's fine to recreate the table. 

------------------------------------
I have tested various flows in the app (variable product creation, editing, order creation with variations in the cart, order detail with variations in the cart, POS) and it all seems to be working as expected. I also tested "migration" - dropping the database after already having some data in there, and I also didn't discover any issues.  (I discovered an unrelated issue on Attributes subscreen and reported it to Linear)

Great job Wojtek and Petros! Thank you for the super clear description and heads-up about changes that might have a higher potential of introducing regressions! 🎖️ 🙇

ParaskP7 commented:
Awesome, thanks for this review and all the extra context you provided as part of it @malinajirka , much appreciated! 🙇 ❤️ 🚀 

> I think this change is fine, but just to double-check my understanding. I believe the remoteProductId between payload and result is always identical, right? Or have you fixed it based on some bug you discovered?

About the `remoteProductId` between `payload` and `result` being identical, I guess so, but I didn't want to introduce a code change, thus I reverted the change as a "code" fix. 😊

FYI: I (indeed) didn't discover any bug.

> Thanks, that's a nice improvement! The wrappers were introduced as part of the introduction of List Management component in WPAndroid (for posts) and later in WCAndroid (for orders). I'm not aware of any guidelines/best-practices regarding the usage - for models that are using the list management component, this is required, for other models, it's nice to have.

Interesting, good to know, didn't have (or forgot) this knowledge about the list management component and all. 🔖 

> I can confirm the byteBuddy exception is happening on my side as well 🤔. Having said that, it's happening even when I run the command on trunk, so I don't think it's related this PR.

Yea, I have the feeling this all started when #13975 merged to `trunk` 3 weeks ago.

> Editing products and variations is supported in the app, but we do not support offline => the user needs to explicitly save the changes while being online => AFAIK it's fine to recreate the table.

Good to know! 👍 

Btw, of all the "legacy" WellSQL tables we still have to migrate, do you maybe have any idea if any such table would require a more careful migration, like migrating its `data` too (due to offline support, etc)? 🤔

---

> I have tested various flows in the app (variable product creation, editing, order creation with variations in the cart, order detail with variations in the cart, POS) and it all seems to be working as expected. I also tested "migration" - dropping the database after already having some data in there, and I also didn't discover any issues. (I discovered an unrelated issue on Attributes subscreen and reported it to Linear)

Thanks for all the testing! 🙇 🧪 🎉 

> Great job Wojtek and Petros! Thank you for the super clear description and heads-up about changes that might have a higher potential of introducing regressions! 🎖️ 🙇

❤️

ParaskP7 committed:
Merge branch 'trunk' of github.com:woocommerce/woocommerce-android into feature/product_variation_to_room

ParaskP7 commented:
👋 @wzieba and just an FYI that I just updated this branch with the latest change from `trunk`. If CI ✅ then we should be ready to merge this. 🚀

Now that you're back, let me know if you want to take one last time at the changes before merging, or want to take it over from here. 🙏

malinajirka commented:
> Btw, of all the "legacy" WellSQL tables we still have to migrate, do you maybe have any idea if any such table would require a more careful migration, like migrating its data too (due to offline support, etc)? 🤔

AFAIK, all our tables are cache only, so they should all be safe to drop - I asked for double confirmation here p1747214664229459-slack-C6H8C3G23.

> Yea, I have the feeling this all started when https://github.com/woocommerce/woocommerce-android/pull/13975 merged to trunk 3 weeks ago.

Ohh, I see 🤔.

ParaskP7 commented:
> AFAIK, all our tables are cache only, so they should all be safe to drop.

Good to know @malinajirka , thanks once more! 🥇

wzieba commented:
> Now that you're back, let me know if you want to take one last time at the changes before merging, or want to take it over from here. 🙏

All fine on my end, thank you for applying improvements!

malinajirka commented:
@ParaskP7 One more thing regarding if it's safe to drop some tables. @wzieba made a good point that SiteModel might potentially cause troubles and I think AccountModel + WCUserModel might as well.

ParaskP7 commented:
> @ParaskP7 One more thing regarding if it's safe to drop some tables. @wzieba made a good point that SiteModel might potentially cause troubles and I think AccountModel + WCUserModel might as well.

Noted @malinajirka , thanks! 🔖 

PS: In the meanwhile, if you think of any other tables that could potentially cause troubles, let us know! 🙏
